plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,403 @@
+				} else if (type instanceof RecordTypeDefinition) {
+					type = ExpressionModelDerivedFeatures.getTypeDefinition(((RecordTypeDefinition)type).getFieldDeclarations().stream().filter(e -> e.equals(fields.remove(0))).findFirst().get().getType());
+				} else {
+					throw new IllegalArgumentException("Type contains forbidden elements: " + type.getClass());
+				}
+			}
+			return ExpressionModelDerivedFeatures.getTypeDefinition(type);
+		}
+		if (expression instanceof FunctionAccessExpression) {
+			Type declarationType = expressionUtil.getDeclaration(expression).getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof RecordAccessExpression) {
+			RecordAccessExpression recordAccessExpression = (RecordAccessExpression) expression;
+			return getType(recordAccessExpression.getFieldReference());
+		}
+		if (expression instanceof SelectExpression) {
+			SelectExpression selectExpression = (SelectExpression) expression;
+			Declaration declaration = expressionUtil.getAccessedDeclaration(selectExpression);
+			TypeDefinition typeDefinition = ExpressionModelDerivedFeatures.getTypeDefinition(declaration);
+			if (typeDefinition instanceof ArrayTypeDefinition) {
+				ArrayTypeDefinition arrayTypeDefinition = (ArrayTypeDefinition) typeDefinition;
+				return arrayTypeDefinition.getElementType();
+			}
+			else if (typeDefinition instanceof IntegerRangeTypeDefinition) {
+				return typeDefinition;
+			}
+			else if (typeDefinition instanceof EnumerationTypeDefinition) {
+				return typeDefinition;
+			}
+			else {
+				throw new IllegalArgumentException("The type of the operand of the select expression is not an enumerable type: " + expressionUtil.getDeclaration(selectExpression));
+			}
+		}
+		/*if (expression == null) {
+			return factory.createVoidTypeDefinition();
+		}*/
+		// EventParameterReferences: they are contained in StatechartModelPackage (they cannot be imported)
+		Optional<EObject> parameter = getParameter(expression);
+		if (parameter.isPresent()) {
+			ParameterDeclaration parameterDeclaration = (ParameterDeclaration) parameter.get();
+			Type declarationType = parameterDeclaration.getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		throw new IllegalArgumentException("Unknown type!");
+	}
+	
+	protected Optional<EObject> getParameter(Expression expression) {
+		return expression.eCrossReferences().stream().filter(it -> it instanceof ParameterDeclaration).findFirst();
+	}
+	
+	// Extension methods
+	
+	// Arithmetics
+	
+	private Type getArithmeticType(Collection<Type> collection) {
+		// Wrong types, not suitable for arithmetic operations
+		if (collection.stream().anyMatch(it -> !isNumber(it))) {
+			throw new IllegalArgumentException("Type is not suitable for arithmetic operations: " + collection);
+		}
+		// All types are numbers
+		if (collection.stream().anyMatch(it -> it instanceof DecimalLiteralExpression)) {
```

Comment:
The collection contains Types. How could any of the elements be of e.g., type DecimalLiteralExpression?